src/python/pants/backend/python/util_rules/pex_requirements.py

```diff
@@ -269,6 +273,74 @@ def __bool__(self) -> bool:
         return bool(self.req_strings)
 
 
+@dataclass(frozen=True)
+class ResolvePexConfig:
+    """Configuration from `[python]` that impacts how the resolve is created."""
+
+    constraints_file: tuple[Digest, FileEntry] | None
+
+    @property
+    def constraints_file_path_and_hash(self) -> tuple[str, str] | None:
+        if self.constraints_file is None:
+            return None
+        file_entry = self.constraints_file[1]
+        return file_entry.path, file_entry.file_digest.fingerprint
+
+
+@dataclass(frozen=True)
+class ResolvePexConfigRequest:
+    """Find all configuration from `[python]` that impacts how the resolve is created."""
+
+    resolve_name: str
+
+
+@rule
+async def determine_resolve_pex_config(
+    request: ResolvePexConfigRequest, python_setup: PythonSetup, union_membership: UnionMembership
+) -> ResolvePexConfig:
+    all_tool_resolve_names = tuple(
+        sentinel.resolve_name for sentinel in union_membership.get(GenerateToolLockfileSentinel)
+    )
```

Comment:
This is weird. It means that our error message for unrecognized resolve names is going to include JVM tools like javac. I could not think of a good way to filter down to only Python tools.

Reply:
Its also a bit weird in the small that you could ask for ResolvePexConfig for the resolve "junit" and get back a non-error answer at all. It definitely seems like a level of typing structure is missing from the infra here.

Reply:
https://github.com/pantsbuild/pants/issues/16443